src/Compilers/CSharp/Portable/Symbols/FileIdentifier.cs

```diff
@@ -9,33 +9,90 @@
 
 namespace Microsoft.CodeAnalysis.CSharp;
 
-internal readonly struct FileIdentifier
+internal struct FileIdentifier
 {
     private static readonly Encoding s_encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
 
-    public string? EncoderFallbackErrorMessage { get; init; }
-    public ImmutableArray<byte> FilePathChecksumOpt { get; init; }
-    public string DisplayFilePath { get; init; }
+    private string? _filePath;
 
-    public static FileIdentifier Create(SyntaxTree tree)
-        => Create(tree.FilePath);
+    private string? _encoderFallbackErrorMessage;
+    private string? _displayFilePath;
+    private ImmutableArray<byte> _filePathChecksumOpt;
 
-    public static FileIdentifier Create(string filePath)
+    private FileIdentifier(string filePath)
+    {
+        _filePath = filePath;
+    }
+
+    private FileIdentifier(ImmutableArray<byte> filePathChecksumOpt, string displayFilePath)
+    {
+        _filePathChecksumOpt = filePathChecksumOpt;
+        _displayFilePath = displayFilePath;
+    }
+
+    private void EnsureInitialized()
     {
-        string? encoderFallbackErrorMessage = null;
-        ImmutableArray<byte> hash = default;
-        try
+        if (_filePath is not null)
```

Comment:
>if (_filePath is not null) [](http://example.com/codeflow?start=8&length=26)

Just realized I'm not sure if the properties are called on different threads. I could put _filePath in a local and use that in the method if that is a concern.

Reply:
Went ahead and changed it, better safe than sorry